testing/sync-test/src/sync15.rs

```diff
@@ -0,0 +1,315 @@
+/* Any copyright is dedicated to the Public Domain.
+http://creativecommons.org/publicdomain/zero/1.0/ */
+//
+// To compile/run from the root directory:
+//
+//     cargo check -p sync-test
+//     cargo run -p sync-test -- --oauth-retries 5
+//
+// (You can safely ignore the noisy 500 for
+// `https://stable.dev.lcip.org/auth/v1/account/destroy` at the end).
+
+use log::*;
+use serde_derive::*;
+use failure::Error;
+use sync_guid::{Guid};
+use sync15_traits::{Store, CollectionRequest, IncomingChangeset, OutgoingChangeset, Payload, ServerTimestamp, StoreSyncAssociation, CollSyncIds};
+use sync15::{telemetry, MemoryCachedState};
+use interrupt::NeverInterrupts;
+use std::cell::{RefCell, Cell};
+use std::borrow::{BorrowMut, Borrow};
+use std::mem;
+
+use crate::auth::TestClient;
+use crate::testing::TestGroup;
+
+// A test record. It has to derive `Serialize` and `Deserialize` (which we import
+// in scope when we do `use serde_derive::*`), so that the `sync15` crate can
+// serialize them to JSON, and parse them from JSON. Deriving `Debug` lets us
+// print it with `{:?}` below.
+#[derive(Debug, Deserialize, Serialize, Clone, Eq, PartialEq)]
+pub struct TestRecord {
+    // This field is required for all Sync records, but can be set to whatever
+    // value we want. In the test, we just generate a random GUID, but we can
+    //  use any GUID we want...
+    // `"recordAAAAAA".into()` also works!
+    pub id: Guid,
+    // To test that syncing happens.
+    pub message: String,
+}
+
+pub struct TestStore {
+    pub name: &'static str,
+    pub test_records: RefCell<Vec<TestRecord>>,
+    pub store_sync_assoc: RefCell<StoreSyncAssociation>,
+    pub was_reset_called: Cell<bool>,
+
+    pub global_id: Option<Guid>,
+    pub coll_id: Option<Guid>,
+}
+
+// Lotsa boilerplate to implement `Store`... 😅
+impl Store for TestStore {
+    fn collection_name(&self) -> std::borrow::Cow<'static, str> {
+        // HACK: We have to use a "well-known" collection name in `meta/global`
+        // (you can see the list in `DEFAULT_ENGINES`, inside
+        // `components/sync15/src/state.rs`). Otherwise, `LocalCollStateMachine`
+        // won't know how to set up the sync IDs. Even though `TestRecord` isn't
+        // actually an address record, that's OK—they're all encrypted, so the
+        // server can't check their contents.
+        "addresses".into()
+    }
+
+    fn apply_incoming(
+        &self,
+        inbound: Vec<IncomingChangeset>,
+        _telem: &mut telemetry::Engine,
+    ) -> Result<OutgoingChangeset, Error> {
+        // Notice the `&mut *` and `.borrow_mut()` to extract the Vec from
+        // the RefCell.
+        let temp: Vec<TestRecord> = mem::take(&mut *self.test_records.borrow_mut());
+
+        let inbound = inbound.into_iter().next().unwrap();
+        for (payload, _timestamp) in inbound.changes {
+            // Here's an example of a magic "into" conversion that we define
```

Comment:
Hmm... I'm not sure this comment is helpful -- saying that we define it ourselves implies to me that these tests are defining it when it's actually a part of the sync15 library.

More broadly, nothing else it says is wrong, but it seems to be trying to explain `into_record`, but if `into_record` is confusing, improving the docs there is probably better than explaining it here...

Is it just the use of HRTB? They're unavoidable with serde, but if so, perhaps a link to something explainign them (rustbook or rust reference) on the into_record doc comment is worth it.